src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,622 @@
+          "Purging files already present on the stage for pipe{} before start."
+              + " reprocessFileSize:{}, files: {}",
+          pipeName,
+          files.size(),
+          String.join(", ", files));
+      conn.purgeStage(stageName, new ArrayList<>(files));
+    } catch (Exception e) {
+      LOGGER.error(
+          "Reprocess cleaner encountered an exception {}:\n{}\n{}",
+          e.getClass(),
+          e.getMessage(),
+          e.getStackTrace());
+    }
+  }
+
+  public static class FileCategorizer {
+    private final Set<String> dirtyFiles = new HashSet<>();
+    private final Map<String, IngestEntry> stageFiles = new HashMap<>();
+    private final long currentOffset;
+
+    static FileCategorizer build(Collection<String> files, long currentOffset) {
+      FileCategorizer categorizer = new FileCategorizer(currentOffset);
+      files.forEach(categorizer::categorizeFile);
+      return categorizer;
+    }
+
+    private FileCategorizer(long startOffset) {
+      this.currentOffset = startOffset;
+    }
+
+    private void categorizeFile(String file) {
+      long fileOffset = FileNameUtils.fileNameToStartOffset(file);
+      long timestamp = FileNameUtils.fileNameToTimeIngested(file);
+      if (currentOffset <= fileOffset) {
+        dirtyFiles.add(file);
+      } else {
+        IngestEntry entry = new IngestEntry(InternalUtils.IngestedFileStatus.NOT_FOUND, timestamp);
+        stageFiles.put(file, entry);
+      }
+    }
+
+    void updateFileStatus(Map<String, InternalUtils.IngestedFileStatus> report) {
+      report.forEach(
+          (fileName, status) ->
+              stageFiles.computeIfPresent(
+                  fileName,
+                  (key, entry) -> {
+                    entry.status = status;
+                    return entry;
+                  }));
+    }
+
+    boolean hasDirtyFiles() {
+      return !dirtyFiles.isEmpty();
+    }
+
+    boolean hasStageFiles() {
+      return !stageFiles.isEmpty();
+    }
+
+    Stream<String> query(Predicate<Map.Entry<String, IngestEntry>> filter) {
+      return stageFiles.entrySet().stream().filter(filter).map(Map.Entry::getKey);
+    }
+
+    void stopTrackingFiles(List<String> files) {
+      files.forEach(
+          file ->
+              stageFiles.computeIfPresent(
+                  file,
+                  (key, entry) -> {
+                    entry.keepTracking = false;
+                    return entry;
+                  }));
+    }
+  }
+
+  @VisibleForTesting
+  static class IngestEntry {
+    private InternalUtils.IngestedFileStatus status;
+    private final long timestamp;
+    private boolean keepTracking = true;
+
+    IngestEntry(InternalUtils.IngestedFileStatus status, long timestamp) {
+      this.status = status;
+      this.timestamp = timestamp;
+    }
+  }
+
+  static class ProgressRegisterImpl implements ProgressRegister {
+    private static final KCLogger LOGGER = new KCLogger(ProgressRegister.class.getName());
```

Comment:
nit: `ProgressRegisterImpl`

Reply:
i didn't want to expose implementation details - what client gets - is a interface reference, thus the logger name here